[DDW-248] Refactor compress/download logs handling

daniloprates opened the pull request:
This PR refactors compress/download logs handling.

## Todo:

- [x] Download logs should always pack fresh logs into `logs.zip` file
- [x] Bug report should always pack fresh logs into `logs.zip` file and submit it to the Report server.
- [x] Create `logs.zip` files with timestamp [1].

[1] Format based in [this answer](https://softwareengineering.stackexchange.com/questions/61683/standard-format-for-using-a-timestamp-as-part-of-a-filename/61748#61748).

## Screenshots:

### Normal state
![image](https://user-images.githubusercontent.com/1504716/42386179-230ad2d4-8115-11e8-87f6-67df01466faf.png)

### Submitting state (Close button is disabled, 50% opacity)
![image](https://user-images.githubusercontent.com/1504716/42386165-1a04f174-8115-11e8-83bd-4c50a05b1d2d.png)


---

## Review Checklist:

### Basics

- [ ] PR is updated to the most recent version of target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance tests are passing (`yarn run test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn run dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn run package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn run flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn run lint`)
- [ ] Text changes are proofread and approved (Jane Wild)
- [ ] There are no missing translations (running `yarn run manage:translations` produces no changes)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn run storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly documented and commented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-rendering
- [ ] Any code that only works in Electron is neatly separated from components

### Testing
- [ ] New feature / change is covered by acceptance tests
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature / change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review:
- [ ] Merge PR
- [ ] Delete source branch
- [ ] Move ticket to `done` on the Youtrack board

nikolaglumac commented on source/common/fileName.js:
```diff
@@ -0,0 +1,7 @@
+import moment from 'moment';
```
@daniloprates please add `// @flow` to the top of this file.

daniloprates replied:
✔︎

nikolaglumac commented on source/common/fileName.js:
```diff
@@ -0,0 +1,7 @@
+  `${prefix}-${moment().format('YYYY-MM-DDThhmmss')}.${filetype}`;
+
+export const getFilenameWithTimestamp = (prefix = 'logs', filetype = 'zip') => fileName =>
+  fileName.match(RegExp(`(${prefix}-)([0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{6})(.${filetype})`));
```
@daniloprates this regexp will have to be updated once you change the filenames to contain timestamp in UTC.

daniloprates replied:
✔︎

nikolaglumac commented on source/main/utils/setupLogging.js:
```diff
@@ -19,4 +21,19 @@ export const setupLogging = () => {
   };
+
+  // Removes existing logs
+  fs.readdir(appLogsFolderPath, (err, files) => {
```
@daniloprates this part works great 🎉

daniloprates replied:
🙂

nikolaglumac commented on source/renderer/app/components/profile/bug-report/BugReportDialog.js:
```diff
@@ -145,18 +147,26 @@ export default class BugReportDialog extends Component<Props, State> {
   componentWillMount() {
-    this.props.onGetLogs();
-    this.props.onDeleteCompressedLogs();
+    this.props.onGetFreshLogs();
```
@daniloprates we need to make sure this call only fetches the log files list and doesn't compress them!

daniloprates replied:
Even when the user submits it?

nikolaglumac commented on source/renderer/app/components/profile/bug-report/BugReportDialog.js:
```diff
@@ -206,24 +216,29 @@ export default class BugReportDialog extends Component<Props, State> {
-          // regular submit
-          this.props.onSubmit(data);
-        }
+        this.props.onSubmit(data);
```
@daniloprates we shouldn't send the log files if the user has set the "attach-logs" switch to off state.

daniloprates replied:
✔︎

nikolaglumac commented on source/renderer/app/containers/profile/dialogs/BugReportDialogContainer.js:
```diff
@@ -61,8 +65,8 @@ export default class BugReportDialogContainer extends Component<InjectedProps> {
         }}
-        onCompressLogs={(logs) => {
-          compressLogs.trigger({ logs });
+        onGetFreshLogs={(logs) => {
```
@daniloprates what is the difference between `onGetLogs` and `onGetFreshLogs`?
If the later one actually compresses the logs then I think the name should be changed...

daniloprates replied:
Renamed it to avoid confusion

nikolaglumac commented on source/renderer/app/stores/ProfileStore.js:
```diff
@@ -224,25 +227,35 @@ export default class SettingsStore extends Store {
     this.isCompressing = true;
-    ipcRenderer.send(COMPRESS_LOGS.REQUEST, toJS(logs));
+    const { fileName = filenameWithTimestamp() } = this.compressedFileDownload;
+
```
@daniloprates we don't need this empty line. Please remove it.

nikolaglumac commented on source/renderer/app/stores/ProfileStore.js:
```diff
@@ -224,25 +227,35 @@ export default class SettingsStore extends Store {
+      fileName: filenameWithTimestamp()
+    };
+    this.isCompressing = true;
+
```
@daniloprates we don't need this empty line. Please remove it.

daniloprates commented:
Generating the logs only upon form submission allowed for a much simpler logic 🙂

nikolaglumac commented on source/common/fileName.js:
```diff
@@ -0,0 +1,8 @@
+import moment from 'moment';
+
+export const filenameWithTimestamp = (prefix:string = 'logs', filetype:string = 'zip') =>
+  `${prefix}-${moment.utc().format('YYYY-MM-DDThhmmss.0SSS')}.${filetype}`;
```
@daniloprates please use the following format `YYYY-MM-DDTHH:mm:ss.0SSS` here (and this should be reflected on the `getFilenameWithTimestamp` function.

nikolaglumac replied:
Oh one thing I forgot - please append `Z` to the end of the timestamp so that it looks like this: `logs-2018-07-04T08:48:19.0182Z.zip`
Thanks! :)

daniloprates replied:
Hey @nikolaglumac!

I'll add the 'Z', however, it's not possible to add the colons, as it's an invalid character for filenames.

daniloprates replied:
✔︎

nikolaglumac commented on source/renderer/app/components/profile/bug-report/BugReportDialog.js:
```diff
@@ -236,6 +236,11 @@ export default class BugReportDialog extends Component<Props, State> {
 
+  onClose = () => {
+    const { error, onCancel } = this.props;
+    if (!this.state.isSubmitting || error) onCancel();
```
@daniloprates closing should be possible in case of an error - we should prevent it only in case of submitting state!

nikolaglumac commented on source/renderer/app/components/profile/bug-report/BugReportDialog.js:
```diff
@@ -302,8 +307,8 @@ export default class BugReportDialog extends Component<Props, State> {
-        onClose={onCancel}
-        closeButton={<DialogCloseButton onClose={onCancel} />}
+        onClose={this.onClose}
+        closeButton={<DialogCloseButton disabled={isSubmitting && !error} onClose={this.onClose} />}
```
@daniloprates closing should be possible in case of an error - we should prevent it only in case of submitting state!

nikolaglumac commented:
@daniloprates can you please post screenshots of how the Dialog close button looks while disabled?
I think @a-rukin should take a look at it ;)

nikolaglumac commented:
@MarcusHurney @DominikGuzei there is one more problem that I have found regarding the Autocomplete component:

**Expected behavior (how it used to be)**
1. Focus the autocomplete component
2. On ENTER key we used to open the dropdown with the list of suggestions as if you typed “a” - or basically the first possible suggestions (the same thing which we do if you click with the mouse in the autocomplete
3. Then if you just keep typing ENTER you will make selections of the first suggestion over again and this is how it should be

**Actual behavior (how it is with the new React-Polymorph)**
1. Focus the autocomplete component
2. On ENTER key the suggestions list are no longer open (nothing happens)
3. If you click with the mouse then the expected top results (starting with “a”) are correctly shown
4. If you type in “b” and select an item and then click again with the mouse the suggestions will show but with results for “b” instead of for top of the list (aka. “a”)

Thanks for fixing it!

daniloprates commented:
@nikolaglumac @a-rukin ✔︎

a-rukin commented:
Close button in disable state - OK

daniloprates commented:
**Commits [27fc619](https://github.com/input-output-hk/daedalus/pull/995/commits/27fc61921d535b0c465c3fb8a2793a10d1c67b2d) and [43a75fc](https://github.com/input-output-hk/daedalus/pull/995/commits/43a75fc04565543384671c437d1ad9dc315ffecd):**

* Moved the `isSubmitting` logic into the `ProfileStore`;
* Renamed `requestFreshCompressedLogs` to `getLogsAndCompress`;
* Renamed `compressedFileDownload.inProgress` to `compressedLogsFile.downloadInProgress`;
* Created `bugReportInProgress`;
* Renamed `getFileNameWithTimestamp` to `isFileNameWithTimestamp`;
* Created `extractFileNameFromPath` for the case we need to get the fileName from the log's path;
